src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/cas2/AssessmentNoteService.kt

```diff
@@ -70,6 +74,39 @@ class ApplicationNoteService(
     )
   }
 
+  @Suppress("ReturnCount")
+  fun createAssessmentNote(assessmentId: UUID, note: NewCas2ApplicationNote):
+    AuthorisableActionResult<ValidatableActionResult<Cas2ApplicationNoteEntity>> {
+    val assessment = assessmentRepository.findByIdOrNull(assessmentId)
+      ?: return AuthorisableActionResult.NotFound()
+
+    val application = applicationRepository.findByIdOrNull(assessment.application.id)
+      ?: return AuthorisableActionResult.NotFound()
+
+    if (application.submittedAt == null) {
+      return AuthorisableActionResult.Success(
+        ValidatableActionResult.GeneralValidationError("This application has not been submitted"),
+      )
+    }
+
+    val isExternalUser = httpAuthService.getCas2AuthenticatedPrincipalOrThrow().isExternalUser()
+    val user = getCas2User(isExternalUser)
+
+    if (!isExternalUser && !nomisUserCanAddNote(application, user as NomisUserEntity)) {
+      return AuthorisableActionResult.Unauthorised()
+    }
+
+    val savedNote = saveNote(application, note.note, user)
```

Comment:
I might have missed something in the docs, but do we not need to start saving the assessment on a note from this point on? So that any new Notes created have an Assessment ID.

Reply:
and we need to start doing it for both `createAssessmentNote` and `createApplicationNote` so that the deprecated end point currently in use starts saving Assessments on a new Note too, and we can safely do the migration to catch the legacy ones

Reply:
Good catch! We duplicate `createAssessmentNote()` but I didn't notice the `saveNote()` actually does the same thing in both cases. Though the commit points to this being a deliberate step I'm not sure why, this does seem to be the critical part.

Reply:
Yes - without this it doesn't work.  I added a note manually yesterday afternoon and expected to see the assessment id being populated but it wasn't.  Then saw assessment was the missing from saveNote() as a parameter.

1. Made me think there is a missing integration test somewhere?  i.e. create a note, then check that the assessment id is returned.  Couldn't see where this test exists or whether we need to create it?

3. Also checking that passing `assessment` to `saveNote()` in requires the following in `AssessmentNoteService.createApplicationNote()`?:
```
val assessment = assessmentRepository.findFirstByApplicationId(applicationId)
      ?: return AuthorisableActionResult.NotFound()
```

Reply:
I've done this in a recent commit and updated the necessary tests.  Hopefully the right way to do it, and will only be used temporarily until the front end is changed (which will be shortly after).